src/main/java/org/opentripplanner/standalone/server/EtagRequestFilter.java

```diff
@@ -0,0 +1,75 @@
+package org.opentripplanner.standalone.server;
+
+import com.google.common.hash.Hashing;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.http.HttpStatus;
+import org.opentripplanner.ext.vectortiles.VectorTilesResource;
+
+public class EtagRequestFilter implements ContainerResponseFilter {
+
+  public static final String DIRECTIVE_NO_STORE = "no-store";
+  public static final String HEADER_ETAG = "ETag";
+  public static final String HEADER_IF_NONE_MATCH = "If-None-Match";
+  public static final String HEADER_CONTENT_TYPE = "Content-Type";
+
+  @Override
+  public void filter(ContainerRequestContext request, ContainerResponseContext response)
+    throws IOException {
+    if (
+      isEligibleForEtag(request, response) &&
+      hasAllowedContentType(response) &&
+      response.getEntity() instanceof byte[] bytes &&
+      bytes.length > 0
+    ) {
+      var clientEtag = request.getHeaderString(HEADER_IF_NONE_MATCH);
+      var etag = generateETagHeaderValue(bytes);
+      var headers = response.getHeaders();
+      headers.add(HEADER_ETAG, etag);
+
+      // if the client's etag matches the generated one then send an empty response
+      if (clientEtag != null && clientEtag.equals(etag)) {
+        response.setEntity(null);
+        response.setEntityStream(new ByteArrayOutputStream());
+        response.setStatus(HttpStatus.SC_NOT_MODIFIED);
+      }
+    }
+  }
+
+  private static boolean isEligibleForEtag(
+    ContainerRequestContext request,
+    ContainerResponseContext response
+  ) {
+    var statusCode = response.getStatus();
+    if (statusCode >= 200 && statusCode < 300 && HttpMethod.GET.matches(request.getMethod())) {
+      String cacheControl = response.getHeaderString("Cache-Control");
+      return (cacheControl == null || !cacheControl.contains(DIRECTIVE_NO_STORE));
```

Comment:
Technically, `ETag`s nothing to do with `Cache-Control`; In fact, the two are complementary: `Cache-Control` tells the browser if (and how long) to cache a resource client-side, whereas `ETag`s "merely" let the browser ask the server if the resource has changed.

In fact, with the [*Does the ETag header make the Cache-Control header obsolete? How to make sure Cache-Control is not harmful then?* StackOverflow question](https://stackoverflow.com/questions/55025948/does-the-etag-header-make-the-cache-control-header-obsolete-how-to-make-sure-ca/55026204#55026204) which shows up in the Google results, I'd argue that [the 2nd answer](https://stackoverflow.com/a/58122775/1072129) is actually the correct (and much more helpful) one:

> Presence of `ETag` header does not tell the browser to do anything. Browser decides what to do based on the `Cache-Control` header it receives in the request and cached response. If it decides that resource is stale or needs to be re-validated, then it can use the `ETag` value to create a conditional request to the server and either get a new resource (status code `200`), or a notification that things have not changed (status code `304`)
>
> Both headers are necessary for your cache to work optimally.

IMO **for vector tiles that *don't* frequently change, we should let the browser cache the tile** for an acceptable period of time; This can be done by setting to e.g. `Cache-Control: max-age=300, stale-while-revalidate=300`, which means:
- the browser can cache the tile client-side and reuse it for subsequent requests for 5 minutes ("fresh"), and
- after the 5 minutes have elapsed, it can reuse the "stale" cached tile for another 5 minutes *while it refetches the tile* from the server.

You can check the [MDN `Cache-Control` page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) for reference on the terms.